### PR TITLE
1168: Use drush_set_error if queue worker callback returns false

### PR DESCRIPTION
**Pull request by bjeavons**
For #1168

### CODE REVIEW

File: commands/core/queue.drush.inc [hunk 1]
```diff
@@ -116,8 +116,11 @@ function drush_queue_run($queue_name) {
   $end = time() + (isset($info['cron']['time']) ? $info['cron']['time'] : 15);
   $queue = drush_queue_get($queue_name);
   while (time() < $end && ($item = $queue->claimItem())) {
-    $function($item->data);
+    $result = $function($item->data);
     $queue->deleteItem($item);
```
Comment (weitzman) [inline commands/core/queue.drush.inc]: Is it right for all use cases to unconditionally delete the item even if $function returned FALSE?

File: commands/core/queue.drush.inc [hunk 2]
```diff
@@ -116,8 +116,11 @@ function drush_queue_run($queue_name) {
...
+    if (isset($result) && $result === FALSE) {
+      drush_set_error('DRUSH_QUEUE_RUN_ERROR', dt('Error during queue run.'));
```
Comment (weitzman) [inline commands/core/queue.drush.inc]: This is going to give the same error for each item that fails. Do we want one error? if not we should ideally state which item is failing. Also, if we rely on the exception technique in D8 we will only get one error message.

File: commands/core/queue.drush.inc (hunk 1 above)
Comment (bjeavons) [inline commands/core/queue.drush.inc]: Worker callback return values are a new concept so I think we get to define what's right. Right now throwing an exception is the only way to keep the item from being deleted, which seems appropriate.

File: commands/core/queue.drush.inc (hunk 2 above)
Comment (bjeavons) [inline commands/core/queue.drush.inc]: There aren't consistent IDs for queued items so I'm not sure how we can report on which item failed. I'd be fine with a single error message tho.

File: commands/core/queue.drush.inc (hunk 1 above)
Comment (weitzman) [inline commands/core/queue.drush.inc]: Makes sense.

File: commands/core/queue.drush.inc (hunk 2 above)
Comment (weitzman) [inline commands/core/queue.drush.inc]: If you get a chance please update the PR, moving this drush_set_error() out of the loop. Probably should be `return drush_set_error()`

**Comment by bjeavons** [comment]
The issue was testQueue isn't expecting a status code of 1. I'll look into the aggregator_feeds run, it may be that this is an actual expected return value.
````
There was 1 failure:
1) QueueCase::testQueue
Unexpected exit code: /home/travis/build/drush-ops/drush/drush --root=/tmp/drush-sandbox/web --uri=dev --nocolor queue-run aggregator_feeds --yes
Failed asserting that 1 matches expected 0.
/home/travis/build/drush-ops/drush/tests/drush_testcase.inc:378
/home/travis/build/drush-ops/drush/tests/drush_testcase.inc:453
/home/travis/build/drush-ops/drush/tests/queueTest.php:37
````

**Comment by weitzman** [comment]
Any progress here, @bjeavons?

**Comment by davereid** [comment]
I don't think we can use this logic unless core does the same.

**Comment by davereid** [comment]
We should be using exception handling only and not a FALSE return logic. There's no standard that queue running callbacks in core should return a FALSE value on failure, so this would likely cause problems with existing queues.

**Comment by davereid** [comment]
I would vote this be closed in favor of https://github.com/drush-ops/drush/pull/1434, which I'm going to work on backporting through the various branches of Drush to ensure it handles exceptions properly.

**Comment by davereid** [comment]
#1434 has been merged, so closing this in favor of the work there.

**Pull request closed without merging**